File: C++/Top 180 Interview questions/solutions/happynumber.cpp
class Solution {
public:
    int func(int n)
    {
        int sum = 0;
        while(n){
        int dig = n % 10;
        n/=10;
        sum+=(dig * dig);
            }
        return sum;
    }
    bool isHappy(int n) {
        unordered_set<int> s;
        while(true)
        {
         if(n==1)
             return true;
         n = func(n);
         if(s.count(n)==1)
             return false;
         s.insert(n);
        }
    }
};

File: Java/HappyNumber.java
class Solution {
    public boolean isHappy(int n) {
        int sum=0;
        do
        {
            do
            {
                int d=n%10;
                sum=d*d+sum;
                n=n/10;
            }while(n>0);
            n=sum;
            sum=0;
        } while(n>9);
        if(n==1)
        {
            return true;
        }
        else
        {
            return false;
        }
    } 
}
